### PR TITLE
KAFKA-3016: phase-1. A local store for join window

### PR DESCRIPTION
**Pull request by ymatsuda**
@guozhangwang 
An implementation of local store for join window. This implementation uses "rolling" of RocksDB instances for timestamp based truncation.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/WindowStoreUtil.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public static final int TIMESTAMP_SIZE = 8;
+    public static final int SEQNUM_SIZE = 4;
+    public static final Serdes<byte[], byte[]> INNER_SERDES = Serdes.withBuiltinTypes("", byte[].class, byte[].class);
+    @SuppressWarnings("unchecked")
+    public static final KeyValueIterator<byte[], byte[]>[] NO_ITERATORS = (KeyValueIterator<byte[], byte[]>[]) new KeyValueIterator[0];
+
+    public static <K> byte[] toBinaryKey(K key, final long timestamp, final int seqnum, Serdes<K, ?> serdes) {
+        byte[] serializedKey = serdes.rawKey(key);
+        byte[] binaryKey = new byte[serializedKey.length + TIMESTAMP_SIZE + SEQNUM_SIZE];
+        System.arraycopy(serializedKey, 0, binaryKey, 0, serializedKey.length);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStoreUtil.java]: Could we use ByteBuffer instead of calling arraycopy and manually done putLong and putInt?

File: streams/src/main/java/org/apache/kafka/streams/state/WindowStoreUtil.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            binaryKey[--index] = (byte) (seqnum >>> (8 * i));
+        }
+
+        return binaryKey;
+    }
+
+    public static <K> K keyFromBinaryKey(byte[] binaryKey, Serdes<K, ?> serdes) {
+        byte[] bytes = new byte[binaryKey.length - TIMESTAMP_SIZE - SEQNUM_SIZE];
+
+        System.arraycopy(binaryKey, 0, bytes, 0, bytes.length);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStoreUtil.java:46]: Ditto, could we use ByteBuffer for readings as well?

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        }
+
+        public void destroy() {
+            Utils.delete(dbDir);
+        }
+    }
+
+    private static class RocksDBWindowStoreIterator<V> implements WindowStoreIterator<V> {
+        private final Serdes<?, V> serdes;
+        private final KeyValueIterator<byte[], byte[]>[] iterators;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java]: iterators => iterator

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    private long currentSegmentId = -1L;
+    private int seqnum = 0;
+
+    public RocksDBWindowStore(String name, long windowBefore, long windowAfter, long retentionPeriod, int numSegments, Serdes<K, V> serdes) {
+        this.name = name;
+        this.windowBefore = windowBefore;
+        this.windowAfter = windowAfter;
+
+        // The retention period must be at least four times as long as the windowSize
+        if ((this.windowBefore + this.windowAfter + 1) * 2 > retentionPeriod)
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:114]: Not clear what do windowBefore / windowAfter mean here? From the code I saw they are set as windowSize or 0 but still not able to figure out the meaning.

EDIT: OK I think now I understand windowBefore and windowAfter, but the comment is not consistent with the code description, should we modify it to "two times the beforeWindowSize and afterWindowSize"? Also maybe rename these two parameters accordingly to make them more understandable.

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            retentionPeriod = (this.windowBefore + this.windowAfter + 1) * 2;
+
+        // The segment interval must be greater than MIN_SEGMENT_INTERVAL
+        this.segmentInterval = Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:118]: Also not clear why "numSegments - 1" here.

Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:118]: Since one of the segments is always empty when it is created, we have to cover the retention period with (numSegment - 1) segments in order to guarantee the retention period.

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        long segmentId = segmentId(timestamp);
+
+        if (segmentId > currentSegmentId) {
+            // A new segment will be created. Clean up old segments first.
+            currentSegmentId = segmentId;
+            cleanup();
+        }
+
+        // If the record is within the retention period, put it in the store.
+        if (segmentId > currentSegmentId - segments.length) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:176]: Not clear about this condition (and the one in line 197), could you elaborate a bit? Specifically, if line 169 passes, then currentSegmentId == segmentId right?

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+            segments[index] = new Segment(name + "-" + directorySuffix(segmentId), segmentId);
+            segments[index].init(context);
+        }
+
+        return segments[index];
+    }
+
+    private void cleanup() {
+        for (int i = 0; i < segments.length; i++) {
+            if (segments[i] != null && segments[i].id <= currentSegmentId - segments.length) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:255]: Ditto here about the condition "segments[i].id <= currentSegmentId - segments.length", could you explain?

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java (hunk 6 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:176]: `currentSegmentId - segments.length` is one less than the smallest valid segment id.

`segmentId` can be less than `currentSegmentId`. We do not assume that records come in the chronological order. The range check is still necessary.

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java (hunk 7 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:255]: `currentSegmentId - segments.length` is one less than the smallest valid segment id.
If the segment id is equal of less than `currentSegmentId - segments.length`, the segment no longer overlaps with the retention period.

File: streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * under the License.
+ */
+
+package org.apache.kafka.streams.state;
+
+import org.apache.kafka.streams.processor.StateStore;
+
+public interface WindowStore<K, V> extends StateStore {
+
+    void put(K key, V value);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java:26]: The original motivation of having the MeteredXXStore is to wrap all metrics / logging semantics into one place so they do not need to be re-implemented again, but this seems to be an obstacle with the current pattern now, for example `MeteredWindowStore.putAndReturnInternalKey` is only used for logging, and `MeteredWindowStore.putInternal` / `MeteredWindowStore.getInternal` are never used since only its `inner` will trigger this function. So how about refactoring this piece as follows:
1. `WindowStore` only expose two APIs: `put(K, V)` and `get(K, long)`.
2. Add a `RollingRocksDBStores` that does not extend any interface, but only implements `putInternal`, `getInternal` and `putAndReturnInternalKey` that uses underlying `RocksDBStore<byte[], byte[]>` as `Segments`.
3. `RocksDBWindowStore` implements `WindowStore` with an `RollingRocksDBStores` inner.
4. Let `MeteredXXStore` only maintain the metrics recording logic, and let different stores implement their own logging logic, since this is now different across different types and are better handled separately. Also some types of stores may not even have a `loggingEnabled` flag, if it will always log, or will never log.

Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java:26]: Can we do it in a separate PR?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java:26]: Sure, could you file a JIRA just to keep track?

Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowStore.java:26]: filed.

File: streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+                segments[i].close();
+                segments[i].destroy();
+                segments[i] = null;
+            }
+        }
+    }
+
+    public Set<Long> segmentIds() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java:272]: nit: Add comment that this is for test only?

File: streams/src/test/java/org/apache/kafka/streams/state/RocksDBWindowStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+                assertEquals(Utils.mkList("two+3", "two+4", "two+5", "two+6"), toList(store.fetch(2, startTime + 8L)));
+                assertEquals(Utils.mkList("two+4", "two+5", "two+6"), toList(store.fetch(2, startTime + 9L)));
+                assertEquals(Utils.mkList("two+5", "two+6"), toList(store.fetch(2, startTime + 10L)));
+                assertEquals(Utils.mkList("two+6"), toList(store.fetch(2, startTime + 11L)));
+                assertEquals(Utils.mkList(), toList(store.fetch(2, startTime + 12L)));
+
+                // Flush the store and verify all current entries were properly flushed ...
+                store.flush();
+
+                Map<Integer, Set<String>> entriesByKey = new HashMap<>();
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/RocksDBWindowStoreTest.java]: nit: can we put this piece of code as a private function that can be shared among three test cases?

**Comment by guozhangwang** [comment]
Some minor comment left, otherwise LGTM. Waiting for Jenkins results.

File: streams/src/test/java/org/apache/kafka/streams/state/RocksDBWindowStoreTest.java (hunk 10 above)
Comment (ymatsuda) [inline streams/src/test/java/org/apache/kafka/streams/state/RocksDBWindowStoreTest.java]: ok
